Update README.md

HitCache opened the pull request:
Minor but important spelling fix.

Leopotam commented on README.md:
````diff
@@ -414,4 +414,4 @@ class TestWorldEventListener : IEcsWorldEventListener {
 world.AddEventListener (listener);
-``` 
\ No newline at end of file
+``` 
````
Can you revert this line, please? Its not related to mistype above.

HitCache replied:
To be honest I'm not sure how to revert that, I did the change quickly in GitHub online and the deletion/re-add of line 417 happened automatically from GitHub's side.

Leopotam reviewed (changes requested):
Can you fork from `develop` branch instead of `master`? `master` branch uses for releases, `develop` - for all changes, fixes, etc.

HitCache commented:
Changed the target fork to be develop, attempted to fix the line thing but that's noise introduced on GitHub's side I really don't know how to fix that, sorry :( I used their client to edit the readme and all I changed was that one typo.

I know it's undesirable but hopefully that's ok. It's a very small change in general I can just throw away the pull request if you're fine to fix the typo from your end, doesn't matter much to me.

Leopotam commented:
@HitCache ok, I fixed issue on my side, thanks for report!

ApmeM commented:
Я почитал каменты к другим пул реквестам, и понял, что обратная совместимость тут так же важна (что вполне понятно). В этом пул реквесте на самом деле обратная совместимость нарушена.

Я могу оформить это изменение так же с флагом компиляции как и другие флаги, но возникает вопрос - видите ли вы смысл в данном изменении и надо ли мне этим в принципе заниматься? Я могу для себя это изменение оставить и периодами смерживаться как из апстрима и меня это вполне устраивает.

Leopotam commented:
Посмотрел, включая прочие "бенчмарки". Не согласен, по пунктам, от общего к частному:
1. ECS - это не про скорость, а про архитектуру. Про скорость - это DOD. Юнитехи мозги хорошо промыли всем, смешав эти 2 разных понятия до отсутствия разницы - тот же Entitas по таким измерениям в принципе не должен существовать в наши дни, но это самый популярный фреймворк с самым жирным сообществом. Иногда требуется возможность регистрации существующего класса в качестве системы. Интерфейсы позволяют менее инвазивно производить интеграцию.

2. Предлагается прогонять Run() методы у всех систем. В реальных проектах их может быть порядка 400 штук в одном стартапе (и это не максимум), из них как минимум треть - это чистые Init/Destroy-системы, но они все-равно будут гоняться, пусть и с пустыми методами. Почему не фильтруешь нужные системы в отдельный кеш `_runSystems` как было раньше (но уже не по интерфейсу, а базовому классу) - хз, это позволит игнорировать нагрузку систем, отличных от Run/PostRun.

3. Бенчи в первом приближении некорректны:
3.1. Если меряешь инициализацию на создание кучи сущностей - должен прокинуть в конструктор мира конфиг с нужным количеством сущностей изначально, иначе ты просто меряешь скорость Array.Resize().
3.2. Если меряешь Run-системы, то не должен собирать фильтры и звать пулы в цикле - это медленно и про необходимость кеширования написано в README. Кеширование можно сделать в поля системы через Init-систему или через `ecslite-di`-инъекцию.
3.3. Судя по результатам произошел инлайн Run() метода системы, чего на реальных проектах никогда не будет просто исходя из количества логики в нем. Надо добавить атрибут запрета инлайна для более корректного сравнения. Ну и это, скорее всего, основная причина разрыва по циферкам - интерфейсные методы не могут быть заинлайнены.
3.4. Все тесты сильно зависят от рантайма. В том же il2cpp результаты могут быть другими.

Я согласен, что вызов виртуального метода быстрее, чем вызов интерфейсного метода, но на реальных проектах разница будет невелика на фоне остальной нагрузки логикой систем, тем более без инлайна.

Юнит-тесты есть, но только внутренние, в пабличный доступ не отдаются, т.к их придется поддерживать.

Делать ломающих изменений в ядро пока не планирую - подобные штуки поломают все расширения и весь существующий код всех проектов. Да и смысла большого нет, ядро позволяет использовать кастомные реализации `IProtoSystems` - можно сделать расширение со своим `FastProtoSystems`, реализующим интерфейс `IProtoSystems` и требованием базового класса у систем - пользователи так уже делают для всяких специальных штук типа временного отключения блока систем или систем с непостоянной частотой вызова.

ApmeM commented:
1. С этим спорить не буду. Просто я сам не очень люблю наследование, и стараюсь его максимально избегать. Но да пожалуй это основная причина почему мое изменение на подойдет во всех случаях. (Предложеный вариант формально конечно тоже наследование, но "это другое" :) )
2. При условии, что бенчмарк верен (см п.3), пустые циклы в одной трети реальных систем всеравно будут быстрее.
3. А вот это надо проверить:
3.1. Перенесу создарие сущностей в еще одну систему, чтобы они создавались каждый тик еще по одной.
3.2. В этом бенчмарке это сделано преднамеряно, чтобы тоде оценить скорость. Конечно для совсем правильной оценки стоит делать кучу нещависимых замеров, но в первую очередь было желание оценить именно свое измерение на фоне всех действий, а для этого в общем то не важно как получаются фильтры  т.к. с ними и не связано (как в общем то и создание ентитей из п.3.1.)
3.3. Запрещу инлайн. Если это оно, то сразу видно будет.

А реально у меня просто нет большого опыта в использовании ецс, потому возможно и делаю ерунду.

Leopotam commented:
> При условии, что бенчмарк верен (см п.3), пустые циклы в одной трети реальных систем всеравно будут быстрее.

Надо пометить вызов Run() в системе `[MethodImpl (MethodImplOptions.NoInlining)]` и перемерять еще раз. Пустые методы могут быть заинлайнены и скорость не пострадает, но это зависит от рантайма в общем случае. il2cpp, к сожалению, является основным рантаймом, на который приходится ориентироваться. Официальный дотнет мало используется пользователями лайта, если только на бекенде.

> Перенесу создарие сущностей в еще одну систему, чтобы они создавались каждый тик еще по одной.

Я скорее имел ввиду корректную инициализацию мира с конфигом, в котором мир сразу выделяет место под нужный кеш сущностей:
```c#
var world = new EcsWorld (new EcsWorld.Config { Entities = 10000 });
```

> В этом бенчмарке это сделано преднамеряно, чтобы тоде оценить скорость.

Ну такое, по сути постоянно дергаешь маску из пула, собираешь массив, потом сортируешь его через Array.Sort(), считаешь по аррею хеш, ищешь в дикте существующий фильтр и отправляешь маску в пул. Пулы тоже в диктах ищутся, что тоже не быстро. Это же не тест производительности итераций по фильтру получается, а какой-то микс.

ApmeM commented:
Все хорошо (ну какбы).
Косяк найден, цифры сравнялись.
(пропустил слово override)
Смысла в реквесте больше нет.
